SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUpstreamChecker.java

```diff
@@ -0,0 +1,170 @@
+    final List<UpstreamInfo> extraUpstreams = getExtraUpstreams(upstreamsInBaragonForRequest, upstreamsInSingularityForRequest);
+    LOG.info("Making and sending load balancer request to remove {} extra upstreams.", extraUpstreams.size());
+    LOG.info("Upstreams removed are: {}", Arrays.toString(extraUpstreams.toArray()));
+    return lbClient.makeAndSendLoadBalancerRequest(loadBalancerRequestId, Collections.emptyList(), extraUpstreams, deploy, singularityRequest);
+  }
+
+  private boolean noPendingDeploy() {
+    return deployManager.getPendingDeploys().size() == 0;
+  }
+
+  private void doSyncUpstreamsForService(SingularityRequest singularityRequest) {
+    if (singularityRequest.isLoadBalanced() && noPendingDeploy()) {
+      final String singularityRequestId = singularityRequest.getId();
+      LOG.info("Doing syncing of upstreams for service: {}.", singularityRequestId);
+      final Optional<String> maybeDeployId = deployManager.getInUseDeployId(singularityRequestId);
+      if (maybeDeployId.isPresent()) {
+        final String deployId = maybeDeployId.get();
+        final Optional<SingularityDeploy> maybeDeploy = deployManager.getDeploy(singularityRequestId, deployId);
+        if (maybeDeploy.isPresent()) {
+          final SingularityDeploy deploy = maybeDeploy.get();
+          final Optional<String> loadBalancerUpstreamGroup = deploy.getLoadBalancerUpstreamGroup();
+          final SingularityLoadBalancerUpdate syncUpstreamsUpdate;
+          try {
+            syncUpstreamsUpdate = syncUpstreamsForService(singularityRequest, deploy, loadBalancerUpstreamGroup);
+            final LoadBalancerRequestId loadBalancerRequestId = syncUpstreamsUpdate.getLoadBalancerRequestId();
+            checkSyncUpstreamsState(loadBalancerRequestId, singularityRequestId);
+          } catch (Exception e) {
+            LOG.error("Could not sync upstreams for service. singularityRequestId: {}, deployId: {}.", singularityRequestId, deployId, e);
+          }
+        }
+      }
+    }
+  }
+
+  private void checkSyncUpstreamsState(LoadBalancerRequestId loadBalancerRequestId, String singularityRequestId) {
+    Retryer<BaragonRequestState> syncingRetryer = RetryerBuilder.<BaragonRequestState>newBuilder()
+        .retryIfException()
+        .withWaitStrategy(WaitStrategies.fixedWait(1, TimeUnit.SECONDS))
+        .retryIfResult(IS_WAITING_STATE)
+        .build();
+    try {
+      BaragonRequestState syncUpstreamsState = syncingRetryer.call(() -> lbClient.getState(loadBalancerRequestId).getLoadBalancerState());
+      if (syncUpstreamsState == BaragonRequestState.SUCCESS){
+        LOG.info("Syncing upstreams for singularity request {} is {}.", singularityRequestId, lbClient.getState(loadBalancerRequestId).toString());
+      } else {
+        LOG.error("Syncing upstreams for singularity request {} is {}.", singularityRequestId, lbClient.getState(loadBalancerRequestId).toString());
```

Comment:
for this error case we will want to make sure the `message` from the full `SingularityLoadBalancerUpdate` (from `lbClient.getState`) is logged. That will contain the actual reason for why the update failed if we need it for debugging. So you'll need to just change the retryer to return the full object and the predicate to check the field